src/main/java/org/deepsymmetry/beatlink/AnnouncementSocketConnection.java

```diff
@@ -0,0 +1,294 @@
+                }
+            }, "beat-link DeviceFinder receiver");
+            receiver.setDaemon(true);
+            receiver.start();
+        }
+    }
+
+    public void send(DatagramPacket packet) throws IOException {
+        announcementSocket.get().send(packet);
+    }
+
+    @Override
+    public SocketSender getSocketSender() {
+        return this;
+    }
+
+
+    /**
+     * Keeps track of the registered device announcement listeners.
+     */
+    private final Set<AnnouncementListener> deviceListeners =
+            Collections.newSetFromMap(new ConcurrentHashMap<AnnouncementListener, Boolean>());
+
+    /**
+     * Adds the specified device announcement listener to receive device announcements when DJ Link devices
+     * are found on or leave the network. If {@code listener} is {@code null} or already present in the list
+     * of registered listeners, no exception is thrown and no action is performed.
+     *
+     * <p>Device announcements are delivered to listeners on the
+     * <a href="https://docs.oracle.com/javase/tutorial/uiswing/concurrency/dispatch.html">Event Dispatch thread</a>,
+     * so it is fine to interact with user interface objects in listener methods. Any code in the listener method
+     * must finish quickly, or unhandled events will back up and the user interface will be come unresponsive.</p>
+     *
+     * @param listener the device announcement listener to add
+     */
+    public void addDeviceAnnouncementListener(AnnouncementListener listener) {
+        if (listener != null) {
+            deviceListeners.add(listener);
+        }
+    }
+
+    /**
+     * Removes the specified device announcement listener so that it no longer receives device announcements when
+     * DJ Link devices are found on or leave the network. If {@code listener} is {@code null} or not present
+     * in the list of registered listeners, no exception is thrown and no action is performed.
+     *
+     * @param listener the device announcement listener to remove
+     */
+    public void removeDeviceAnnouncementListener(AnnouncementListener listener) {
+        if (listener != null) {
+            deviceListeners.remove(listener);
+        }
+    }
+
+    /**
+     * Get the set of device announcement listeners that are currently registered.
+     *
+     * @return the currently registered device announcement listeners
+     */
+    @SuppressWarnings("WeakerAccess")
+    public Set<AnnouncementListener> getDeviceAnnouncementListeners() {
+        // Make a copy so callers get an immutable snapshot of the current state.
+        return Collections.unmodifiableSet(new HashSet<AnnouncementListener>(deviceListeners));
+    }
+
+    /**
+     * Stop listening for device announcements. Also discard any announcements which had been received, and
+     * notify any registered listeners that those devices have been lost.
+     */
+    @SuppressWarnings("WeakerAccess")
+    public synchronized void stopAnnouncementSocket() {
```

Comment:
Why not just `stop()`? You are  calling it on an `AnnouncementSocketConnection` after all.

Reply:
And… actually, you have the `stop()` method a few lines further down, so this copy is redundant and can just be deleted.

Reply:
I have removed close and reverted all back to stop.